Fancy  screenshot grabber

Papierkorb opened the pull request:
Took TheSpiritXIII's work and rewrote the grabber to use QGraphicsView instead of QRubberBand, allowing to have fancy features like better visual feedback and being able to resize and move the selection further.

TheSpiritXIII committed:
Screenshot file transfer and smiley icon change

TheSpiritXIII committed:
Fix conventions, icon/button regreesion

TheSpiritXIII committed:
Merge remote-tracking branch 'upstream/master'

TheSpiritXIII committed:
Selection can be made anywhere, improved blocking interface

Papierkorb committed:
Merge branch 'master' of github.com:TheSpiritXIII/qTox

Papierkorb committed:
Screenshot: Use a QGraphicsView as drawing back-end

This will enable us to easily add more complex functionality later on.
Work in progress, more to come.

Papierkorb committed:
Screen grabber: Accept selected region on double click

Papierkorb committed:
Screenshot grabber: Added helper tooltip center-top of screen

Papierkorb committed:
Screen grabber: Don't draw black above chosen rect

Papierkorb committed:
Screen grabber: Added handles to further resize the selection

Papierkorb committed:
Screen grabber: Removed left over debug output

zetok commented:
```
 * Applying user patches from /etc/portage/patches//net-im/qtox ...
 *   1332-zetok-quoting-fix.patch ...                                                                                                        [ ok ]
 *   1371-zero-one-alphabetized-friend-list.patch ...                                                                                        [ ok ]
 *   1462-Papierkorb-screenshot-button.patch ...                                                                                             [ ok ]
 * Done with patching
```

![>tfw no screenshot](https://cloud.githubusercontent.com/assets/3148759/6810082/6cd26826-d256-11e4-97a6-a419fb4e9be6.png)

How one is supposed to take that screenshot? → Could you please use idea that makes it clear, since obviously currently it's **not** intuitive, and you're _hiding_ functionality that should be somewhat exposed to user.

Colour selection is _bad_. E.g. nicely done colour in ksnapshot: 
![ksnapshot nice](https://cloud.githubusercontent.com/assets/3148759/6810487/dd7d57dc-d258-11e4-8bcc-6fed604b3264.png)

And qTox:
![qTox bad](https://cloud.githubusercontent.com/assets/3148759/6810497/e95f79d6-d258-11e4-8ed6-8b7ba25b1f58.png)

Also, I don't think that including commits with binary blobs (.png) that aren't used anyway is good…

Beside that, it would be nice if you could put https://github.com/tux3/qTox/wiki/Coding#coding-style into practice, since currently PR has mixed style..

zetok commented:
Oh, and I forgot to mention; selection performance is bad.

Papierkorb commented:
> How one is supposed to take that screenshot?

Granted, I focused on implementing the feature, and less on exposing it nicely. The current context-menu is a sub-par solution. I was thinking about replacing the current "attach" button functionality with that it simply shows a popout menu, where the user can choose "Attach file" and "Attach screenshot". Not sure if this is ideal, but certainly better I think. And considering that most users probably drag-drop a file into the window to send it, it may not even introduce a single extra click for most users?

The PNG is an left over of the PR I based this upon. Will throw it out when we decided how to expose this feature, that is, either do what I propose above, or have the "dropout" arrow next to the attach button. (I'm fine with doing it, but as a user, I don't like it when the UI is cramped full with those downwards arrows)

> Colour selection is bad.

What would be a better choice? Some lighter blue color like KSnapshot?

> selection performance is bad.

Really? Could you elaborate what's lagging or "feels bad"? Does it lag when you create the selection? Or move it, or resize it using the handles? Could you provide some HW details? Works fine on my notebook with i5 520M CPU/GPU.

> Beside that, it would be nice if you could put https://github.com/tux3/qTox/wiki/Coding#coding-style into practice, since currently PR has mixed style..

Will have a look, thanks.

ProMcTagonist commented:
@Papierkorb What does a "popout" menu look like?

Papierkorb commented:
Note: Just noticed I forgot to swap out the license header in my IDE. Sorry. Will also fix this right away.

Papierkorb committed:
Screen grabber: Fixed license headers and code style

Hope I didn't overlook anything. Super sorry about the license headers.
No idea if I'd should throw my PR away and recreate it to purge my
license headers from the qTox git history.

Apart from that, moved the pointer star to the Type* , removed spaces
between function names and (), formatted one-line if bodies to not
use braces, and moved the opening brace of multi-line if bodies to
their own line.

Papierkorb commented:
@ProMcTagonist  A popout menu would look like this (It's effectively a context-menu appearing when you click the button):
![menu_right](https://cloud.githubusercontent.com/assets/737062/6812099/f17ac620-d26c-11e4-993c-599652d4c75a.png)
or this
![menu_bottom](https://cloud.githubusercontent.com/assets/737062/6812102/f95f59be-d26c-11e4-8608-34b69a97bb1c.png)

zetok commented:
> Or move it, or resize it using the handles?

Yep.
Flamegraph: https://a.pomf.se/ppkamb.svg

ksnapshot's selection works just fine, but it uses Qt 4, so..

```
Kernel: 3.18.7-gentoo
Xorg: 1.16.4
nouveau: 1.0.11
CPU: AMD Phenom II 550 BE @ 3.6 GHz
GPU: nvidia gtx 460
```

# 

> A popout menu

No.

# 

You probably would want to implement button according to #773, since it seems to be superior to what you propose. (if you want for me to elaborate on that, just ask (though don't expect quick answer))

Papierkorb commented:
@zetok Okay, it spends most of its time for drawing the QPixmap (which is the screen capture). Currently, most of the screen will be redrawn on each change. I'll look into it to make it draw less, and then I can just hope that Qt will do some trick and only draw the pixmap partially.

The button. So, I hover (or click?) over the attach button, and a list of buttons (like a toolbar) expands to the left, showing buttons for e.g. screen capture?

zetok commented:
> The button. So, I hover (or click?) over the attach button, and a list of buttons (like a toolbar) expands to the left, showing buttons for e.g. screen capture?

Yep, hover. Note that just clicking on attach button remains a default action, i.e. it sends file.

ProMcTagonist commented:
I like #773 but they should be shaped like this

```
/__|__|__|__/
```

instead of how they're shown in the first post in #773 which is

```
|__/|__/|__/|__/
```

Can't fire up GIMP right now, hope that makes sense.

zetok commented:
If I understood correctly, `|__/|__/|__/|__/` refers to corners, right?
Um, you know, #773 was just a quick copypaste made in gimp, so…  `;)`

What you propose seems good to me :3

ProMcTagonist commented:
Yes, exactly.

Cool.

Papierkorb commented:
Like this?
![screenshot_slider](https://cloud.githubusercontent.com/assets/737062/6813118/5d615e3e-d273-11e4-8d37-99c26e9f9ea4.png)

ProMcTagonist commented:
Flip the monitor one vertically.

Papierkorb commented:
Alternative versions: 

First:
![screenshot_slider_edge](https://cloud.githubusercontent.com/assets/737062/6813166/be06f7e4-d273-11e4-917d-5028f79d4df9.png)

Second: (My favorite)
![screenshot_slider_up](https://cloud.githubusercontent.com/assets/737062/6813165/be04c8b6-d273-11e4-95e9-ecf208a34009.png)

ProMcTagonist commented:
Second :+1:

ghost commented:
@Papierkorb Can you make it slide out on mouse hover?

Papierkorb commented:
@mannol That's the plan :)

Progress: Created the button SVG based on my GIMP mockup.

ghost commented:
@Papierkorb Oh right, tl;dr.

Papierkorb committed:
Screen grabber: Flyout menu for the screenshot button

When hovering over the "attach file" button in the chat form, an
additional button for the screenshot functionality will 'fly out'
to the left, showing a computer monitor as icon. Leaving the attach
file or the take screenshot button will collapse the fly out again.

Bug: Moving the mouse over the fly out and then back again to the
attach button collapses the fly out. Will sort this out later.

Also used the opportunity to rename headers from hpp -> h extension
I added earlier.

Papierkorb commented:
This took way longer than needed. But it works for the most part. Thoughts?

ProMcTagonist commented:
Does it introduce any new dependencies?

zetok commented:
@ProMcTagonist ?

zetok commented:
https://a.pomf.se/quwzvf.webm

It's awesome, though, can screenshot button be made to not hide after moving cursor 'attach → screenshot → attach' ?

zetok commented:
Also, when making screenshot: 
![qtox-screenshot-ztvu2649a](https://cloud.githubusercontent.com/assets/3148759/6821024/246f8f66-d2cc-11e4-8466-a779650518b6.png)

Papierkorb commented:
Mh, should the flyout collapse first and then take a screenshot (Delays it by ~300ms), or should the screenshot taken while it's out, or should the flyout just disappear?

zetok commented:
I don't know, but halfway through isn't really good

zetok commented on src/widget/form/chatform.cpp:
```diff
@@ -879,6 +884,37 @@ void ChatForm::loadHistory(QDateTime since, bool processUndelivered)
 
+void ChatForm::onScreenshotClicked()
+{
+    ScreenshotGrabber *screenshotGrabber = new ScreenshotGrabber (this);
```
space ()

Papierkorb replied:
Will be fixed in the now upcoming commit

Papierkorb committed:
Screen grabber: Wait for the flyout to collapse before grabbing

After clicking the "Screenshot" button in the file flyout, the flyout
now collapses and then triggers the screenshot functionality. This adds
a delay of ca. 1/3 seconds between the click and the action.

Papierkorb committed:
Screen grabber: Moving from button -> flyout -> button doesn't collapse

Papierkorb commented:
Ok, the flyout should not be visible at all anymore on screenshots, and doing the button <-> flyout dance should not collapse it either. If everythings fine, I'd take look at the performance stuff now.

@ProMcTagonist No additional dependencies. Pure Qt.

zetok commented:
> This adds a delay of ca. 1/3

Adding a delay is kinda worse than just a "glitch" of halfway hidden button

Papierkorb commented:
Well, does it feel bad? Before that commit, there was already a delay which was a left over of the context-menu stuff. Sure we can change it, just want to make sure that this isn't a placebo where something feels worse than before without any real change in timing towards the user.

zetok commented:
> Well, does it feel bad?

Yes.

Papierkorb committed:
Screen grabber: Remove forced redraw

If this doesn't help we'll need to split the screen capture display
into tiles and display those instead, allowing Qt to only draw
a few of them instead of everything.

Papierkorb commented:
Could you check again the screen grabber performance @zetok ? Threw out a forced redraw, but I think we'll need to split the capture into tiles and display those instead ..

Papierkorb committed:
Screen grabber: Don't wait for the flyout to collapse

The menu is now collapsed in the background half a second after
the user clicked on the screenshot button. This is to let the WM have
some time to open the fullscreen window, which itself is now shown
immediately.

zetok commented:
When resizing selection, performance is ~on par (AFAICT with naked eye) with ksnapshot's performance - it's awesomely smooth.

There is still bad performance when dragging selection around.
Flamegraph: https://a.pomf.se/quuihw.svg

Also, there's a graphical glitch when selecting area: https://a.pomf.se/qjjizc.webm

TheLastProject commented on src/widget/tool/screenshotgrabber.cpp:
```diff
@@ -0,0 +1,183 @@
+
+void ScreenshotGrabber::useRegionSelectedTooltip()
+{
+    this->helperTooltip->setHtml(tr("Press <b>Return</b> to accept, or move and resize the selected region.",
```
I don't quite like this text. Return is more commonly known as "Enter", "accept" is vague, there is no text about cancelling and the text implies you can't make a new selection altogether. What about changing it to this:

`Press <b>Enter</b> to send a screenshot of the selected region, <b>Escape</b> to cancel, or select a new region.`

TheLastProject commented on src/widget/tool/screenshotgrabber.cpp:
```diff
@@ -0,0 +1,183 @@
+
+void ScreenshotGrabber::useNothingSelectedTooltip()
+{
+    this->helperTooltip->setHtml(tr("Click and drag to select a region to select it. Press <b>ESC</b> to cancel.",
```
"select a region to select it" is very double. What about simply:

`Click and drag to select a region. Press <b>Escape</b> to cancel.`

Papierkorb replied:
Agreed. I changed the texts to yours with little changes. Ty

Papierkorb committed:
Screen grabber: Modified helper tooltip texts

The english texts are now more coherent

Papierkorb committed:
Screen grabber: Reject small selections (< 2px width/height)

When the user just clicks, or moves the cursor barely (less than 2px),
auto reject the region and show no selection.

Papierkorb committed:
Screen grabber: Show correct tooltip for selected/no-selection states

When the user just clicks somewhere, and thus rejects the selection due
to the previous commit, show the initial 'nothing selected' helper
tooltip again.

Papierkorb committed:
Screen grabber: Calculate bounding rect, fixes some drawing issues

This commit should fix drawing artifacts occuring when quickly clicking
on the screen while moving the cursor in circles.

Papierkorb committed:
Screen grabber: Fix position of handles with negative selection

When you clicked, and then move towards top-left with the cursor,
and release, the handles were displayed inside the resulting rectangle,
instead of outside.

Papierkorb commented:
So, the drawing artifacts should be gone now. Also fixed a off by one pixel error. It's actually surprising, that moving the selection is more costly than resizing, would've expected it to be the other way around...

zetok commented:
So, since PR didn't change in last few days, and last time I've downloaded patches is from `03-28`, there are still bugs with it.

![groupchat screenshot - gl](https://cloud.githubusercontent.com/assets/3148759/6960617/1ac6118c-d91c-11e4-8d97-d4caa1436d16.png)
↑ never-hidding screenshot button in groupchat

I'm not sure how to reproduce whole combination reliably, since it's a bug on top of another bug.

Well, first bug is that it **never hides**, steps to reproduce (haven't tested them a lot):
1. Move cursor between `attach` and `screenshot` buttons, and at some point move cursor outside of qTox window in a way that will leave `screenshot` button visible
2. Minimize qTox to tray by clicking on a tray icon
3. Restore qTox from tray by clicking on tray icon
4. You've just gained never-hiding `screenshot` button! :3 (it would hide after clicking on it, _but who would want to click on it anyway_, right?)

So, the second bug where it **changes its place**.. hmm, I've just been using qTox for some time, and it would seem that for some reason button first moved down, to appear above later on. Not quite sure what it causes, since I've changed chat windows, switched between chat windows and groupchat, opened settings panel..

Both of those bugs are reproducible.

Papierkorb committed:
Merge branch 'master' of github.com:tux3/qTox

Papierkorb committed:
Screen grabber: Only show flyout if the attach button is enabeld

Fixes the bug that the flyout was shown in group chats.

Papierkorb committed:
Screen grabber: hide() the flyout at first

I don't remember why I did show() at first, but works fine without, and
fixes the issue of a 1px wide flyout showing up somewhere in the
chatform

Papierkorb committed:
Screen grabber: Hide mid-animation when leaving file button

The flyout is now more responsive and will start collapsing while it's
expanding, when the user leaves the attach file button mid-animation.

Papierkorb committed:
Screen grabber: Make sure the flyout is hidden on leave

When the user moved from right to left in moderate speed, the flyout
didn't get the mouse leave event.

Papierkorb commented:
Knock knock, would be great if someone would provide some feedback. Thanks

zetok commented:
I still can reproduce never-hiding `screenshot` button. A little bit harder, but.. :3

To reproduce, environment should allow you to close/quit given application using keyboard shortcut, or in any other way that doesn't involve doing anything with mouse.

Steps to reproduce:
1. Set `Close to tray` in settings
2. In contact window move cursor to show `screenshot` button
3. Use keyboard shortcut to close qTox window, which closes qTox to tray
4. Particular contact window now has never-hiding `screenshot` button

Other contacts (unless done separately for them) don't have never-hiding `screenshot` button, and groupchats don't have that problem.

If I may ask you, could your code have screenshot button for groupchats, but be "commented out", so that when groupchats will finally support sending / sharing files, all that would be needed is "uncommenting" that code to make screenshot button work in groupchats?

Papierkorb commented:
Nice catch. I'll just bind it to the hideEvent() of the widget, so if it's hidden, the flyout will be hidden. That should get rid of it once and for all :)

> If I may ask you, could your code have screenshot button for groupchats, but be "commented out", so that when groupchats will finally support sending / sharing files, all that would be needed is "uncommenting" that code to make screenshot button work in groupchats?

That will just work™: The flyout is enabled if the attach file button is enabled. There's a small method in ChatForm which opens the screen grabber, but it just needs to be moved to the GenericChatForm, just like the method used to select and send a file.

I thought about moving that to GenericChatForm, but for that I'd also need to move the "sendFile" signal (No idea why that's a signal) from the ChatForm and the GroupChatForm into the generic class. That'd be a nice thing, but yeah..

Papierkorb committed:
Screen grabber: Hide the flyout when the chat form is hidden.

zetok commented:
I think that I just found a way to reproduce flying screenshot button. 

Steps to reproduce on 181d748c5b5c46a455906acc224515512aaad700:
1. Make it never-hiding
2. Resize qTox's window in i3wm  (not sure if it will "work" in other DE / WM, since i3wm is the only one under which I've tried)
↓
![flying screenshot button](https://cloud.githubusercontent.com/assets/3148759/6994616/317c25bc-db18-11e4-909c-42abb05e2842.png)

Papierkorb committed:
Screen grabber: Fix flying-flyout bug when resizing window while open

Papierkorb commented:
@zetok Fixed :3

Papierkorb commented:
Closing this PR, please see https://github.com/tux3/qTox/pull/1510 instead.
